README.md
# TracebackFilter
Tails a log file and filters for python-style Tracebacks, sending them to an AWS SNS topic

## Usage
```bash
export TBF_SNS_TOPIC="arn-to-your-sns-topic-here"
export TBF_SNS_SUBJECT="Your SNS subject Prefix here"
java -jar tracebackfilter.jar /path/to/your/logfile/here.log
```

## Dockerhub
```bash
docker pull mystickphoenix/tracebackfilter
docker run -d -v /path/to/your/logfile/here.log:/log/monitor.log -e TBF_SNS_TOPIC="arn-to-your-sns-topic-here" -e "Your SNS subject Prefix here" mystickphoenix/tracebackfilter
docker run -d -v /path/to/your/log/folder/here:/log -e TBF_SNS_TOPIC="arn-to-your-sns-topic-here" -e "Your SNS subject Prefix here" mystickphoenix/tracebackfilter /log/your-logfile-name-here.log
```

When specifying the docker args, the -v should map a path to the folder of your logfile.  The two -e env args are required and should be filled out with your specific information.  

The final unnamed argument is the local-to-docker path of the logfile.  So, if in the -v argument you mapped /log/specialfolder:/log and wanted to tail super-secret.log you would then specify /log/super-secret.log as the final argument.

## TODO
* User-defined "fenceposts" for what to extract (regexes?)
* Reading from STDIN
* Writing to STDOUT
* Writing to something other than SNS

## Contributing
Fork + Branch and create a PR, open an issue, or start a conversation!


CHANGELOG.md
# Change Log

## [2016.12.xx][unreleased]
### Changed
- Add environment vars for user configurable fenceposts:
  - `TBF_START_REGEX`
  - `TBF_END_REGEX`
  - `TBF_SKIP_REGEX`
- Reduce the size of the uberjar by removing extra amazonica modules
- Fix subject extraction bug
- Change versioning scheme from semantic to time-based

## [2016.10.27](https://github.com/bryan-lott/TracebackFilter/releases/tag/v2016.10.27)
### Fixed
- Instead of following a single log file endlessly, TF now plays nicely with logrotate and stays on the primary log.
- Bugs squashed

## [2016.10.22](https://github.com/bryan-lott/TracebackFilter/releases/tag/v2016.10.22)
### Initial Release


test/loginfo/core_test.clj
(ns tracebackfilter.core-test
  (:require [clojure.test :refer :all]
            [tracebackfilter.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


test/tracebackfilter/core_test.clj
(ns tracebackfilter.core-test
  (:require [clojure.test :refer :all]
            [tracebackfilter.core :refer :all]))

(def real-log ["Traceback (most recent call last):"
               "  File \"./send_status.py\", line 227, in <module>"
               "smtplib.SMTPDataError: (454, 'Temporary service failure')"
               ""
               "Attempt 1 failed! Trying again in 4 seconds..."
               "Fri Oct 21 12:15:40 UTC 2016 [skipping] report"
               "[skipping] Running report"])

(deftest test-take-to-first
  (testing "Found a split point"
    (is (= [1 1 1 1 2]
           (take-to-first even? [1 1 1 1 2])))
    (is (= [1 1 2]
           (take-to-first even? [1 1 2 1 1])))
    (is (= [2]
           (take-to-first even? [2 1 1]))))
  (testing "No split point"
    (is (= []
           (take-to-first even? [])))
    (is (= [1 1 1]
           (take-to-first even? [1 1 1])))))

(deftest test-drop-to-first
  (testing "Drop point"
    (is (= [2 1 1]
           (drop-to-first even? [1 1 2 1 1])))
    (is (= [2 1 1]
           (drop-to-first even? [2 1 1])))
    (is (= [2]
           (drop-to-first even? [1 1 2]))))
  (testing "No drop point"
    (is (= []
           (drop-to-first even? [1 1 1])))
    (is (= []
           (drop-to-first even? [])))))

(deftest test-partition-inside
  (testing "Start/stop"
    (is (= [[true false] []]
           (partition-inside true? false? [true false])))
    (is (= [[true nil false] []]
           (partition-inside true? false? [nil true nil false nil]))))
  (testing "Multiple start/stop"
    (is (= [[true nil nil false] [true nil false] []]
           (partition-inside true? false? [nil true nil nil false nil nil true nil false nil]))))
  (testing "No start/stop"
    (is (= [[]]
           (partition-inside true? false? [nil nil nil])))
    (is (= nil
           (partition-inside true? false? []))))
  (testing "Start no stop"
    (is (= [[true nil nil] []]
           (partition-inside true? false? [nil true nil nil]))))
  (testing "Nested - these are weird..."
    (is (= [[true nil true] [true]]
           (partition-inside true? false? [nil true nil true])))
    (is (= [[true nil nil true nil false] [true nil false] []]
           (partition-inside true? false? [nil true nil nil true nil false nil false nil])))))

(deftest test-start-capture
  (testing "Start capture"
    (is (true? (start-capture "Traceback"))))
  (testing "Don't start capture"
    (is (false? (start-capture "")))
    (is (false? (start-capture "traceback")))
    (is (false? (start-capture " ")))
    (is (false? (start-capture "other message")))))

(deftest test-end-capture
  (testing "End capture"
    (is (true? (end-capture "didn't start with a space or another keyword"))))
  (testing "Don't end capture"
    (is (false? (end-capture " starting with spaces")))
    (is (false? (end-capture "Traceback")))
    (is (false? (end-capture "smtplib.SMTPDataError: (454, 'Temporary service failure')")))
    (is (false? (end-capture "DETAIL: this is an AWS style traceback with extra redshift detail")))
    (is (false? (end-capture "")))
    (is (false? (end-capture "Attempt 12 failed")))))

(deftest test-extract-lines
  (testing "Traceback Exists"
    (is (= [["Traceback (most recent call last):"
             "  File \"./send_status.py\", line 227, in <module>"
             "smtplib.SMTPDataError: (454, 'Temporary service failure')"
             ""
             "Attempt 1 failed! Trying again in 4 seconds..."
             "Fri Oct 21 12:15:40 UTC 2016 [skipping] report"] []]
           (extract-lines real-log))))
  (testing "No traceback"
    (is (= [[]]
           (extract-lines (map str (range 10))))))
  (testing "Successful Retry"
    (is (= []
           (filter unsuccessful-retry?
                   (butlast
                     (extract-lines ["Traceback (most recent call last):"
                                     "  File \"./send_status.py\", line 227, in <module>"
                                     "smtplib.SMTPDataError: (454, 'Temporary service failure')"
                                     ""
                                     "Attempt 1 failed! Trying again in 4 seconds..."
                                     "Fri Oct 21 12:15:40 UTC 2016 [skipping] report"
                                     "[skipping] Running report"]))))))
  (testing "Unsuccessful Retry"
    (is (= [["Traceback (most recent call last):"
                                     "  File \"./send_status.py\", line 227, in <module>"
                                     "smtplib.SMTPDataError: (454, 'Temporary service failure')"
                                     ""
                                     "Attempt 1 failed and there are no more attempts left!"
                                     "Fri Oct 21 12:15:40 UTC 2016 [skipping] report"]]
           (filter unsuccessful-retry?
                   (butlast
                     (extract-lines ["Traceback (most recent call last):"
                                     "  File \"./send_status.py\", line 227, in <module>"
                                     "smtplib.SMTPDataError: (454, 'Temporary service failure')"
                                     ""
                                     "Attempt 1 failed and there are no more attempts left!"
                                     "Fri Oct 21 12:15:40 UTC 2016 [skipping] report"
                                     "[skipping] Running report"])))))))

(deftest test-unsuccessful-retry
  (testing "Filter this out"
    (is (false? (unsuccessful-retry?
                  ["Traceback"
                   "Other stuff Here"
                   "Attempt 1 failed! Trying again in 4 seconds..."
                   ""]))))
  (testing "Don't filter this one"
    (is (true? (unsuccessful-retry?
                 ["Traceback"
                  "Other stuff here"
                  "Attempt 1 failed and there are no more attempts left!"])))))

(deftest test-traceback-type
  (testing "Extract traceback type"
    (is (= "traceback.Thing: "
           (traceback-type ["1" "2" "3" "traceback.Thing: "]))))
  (testing "No traceback found"
    (is (= nil
           (traceback-type ["1" "2"])))))

(deftest test-subject
  (testing "Building traceback subject"
    (is (= "Subject - traceback.Thing: "
           (subject "Subject" ["1" "2" "3" "traceback.Thing: "]))))
  (testing "No traceback found"
    (is (= "Subject - "
           (subject "Subject" ["1" "2"])))))


project.clj
(defproject tracebackfilter "0.1.0-SNAPSHOT"
  :description "extract multi-line log information and send to external"
  :url "http://example.com/FIXME"
  :dependencies [[org.clojure/clojure "1.8.0"]
                 [org.clojure/tools.logging "0.3.1"] ;; logging
                 [org.clojure/tools.cli "0.3.5"] ;; arg parsing
                 [amazonica "0.3.76" :exclusions [com.amaonaws/aws-java-sdk
                                                  com.amazonaws/amazon-kinesis-client]]  ;; talk to aws, exclude everything
                 [com.amazonaws/aws-java-sdk-core "1.10.49"] ;; include aws core
                 [com.amazonaws/aws-java-sdk-sns "1.10.49"] ;; include sns
                 [clj-time "0.13.0"]] ;; timestamps
                                                                                    
  :target-path "target/%s"
  :main ^:skip-aot tracebackfilter.core
  :profiles {:uberjar {:aot :all
                       :jvm-opts ["-Xms4g" "-Xmx4g" "-server"]}})




src/tracebackfilter/core.clj
(ns tracebackfilter.core
  (:import (java.io RandomAccessFile))
  (:require [clojure.string :refer [join starts-with? split]]
            [amazonica.aws.sns :as sns]
            [clojure.tools.cli :refer [parse-opts]]
            [clojure.tools.logging :as log]
            [clj-time.core :as -time]
            [clj-time.format :as format-time])
  (:gen-class))


;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Environment Variable Configs
(defn slack-sns-topic []
  (System/getenv "TBF_SNS_TOPIC"))

(defn slack-sns-subject []
  (System/getenv "TBF_SNS_SUBJECT"))

(defn start-fencepost []
  (or (System/getenv "TBF_START_REGEX")
      #"^Traceback"))

(defn end-fencepost []
  (or (System/getenv "TBF_END_REGEX")
      #"^\s+|^Traceback|^[\w\.]+:\s+|^DETAIL|^$|^Attempt \d+ failed"))

(defn skip-regex []
  "This regex is applied after a section has been captured.  If the regex
   exists, filter *out* the section."
  (or (System/getenv "TBF_SKIP_REGEX")
      #"Attempt \d+ failed! Trying again in \d+ seconds..."))

;;;;;;;;;;;;;;;;;;;;;
;; Sequence Functions

(defn drop-to-first
  "Returns a lazy sequence of successive items from coll after
  the point and including the point at which (pred item) returns true.
  pred must be free of side-effects."
  [before pred coll]
  (lazy-seq
    (when-let [s (seq coll)]
      (if-not (pred (nth s before nil))  ;; acts as a "lookahead" for retaining the before lines
        (drop-to-first before pred (rest s))  ;; move one ahead
        s))))

(defn take-to-first
  "Returns a lazy sequence of successive items from coll up to
  and including the point at which it (pred item) returns true.
  pred must be free of side-effects.
  taken from: https://groups.google.com/forum/#!topic/clojure/Gs6UtrRSLv8"
  [offset after pred coll]
  (when-let [s (seq coll)]
    (if (> offset 0)
      (let [o (take offset s)
            s (drop offset s)
            t (first s)]
        (if-not (pred t)
          (concat o [t] (take-to-first 0 after pred (rest s)))
          (take (inc after) s)))
      (let [t (first s)]
        (if-not (pred t)
          (concat [t] (take-to-first 0 after pred (rest s)))
          (take (inc after) s))))))


(defn partition-inside
  "Grabs whats between and including the pred-start and pred-stop.
  pred-start and pred-stop should not be the same predicate.
  if the data is of a 'nested' format each of the nestings will be returned."
  [before after pred-start pred-stop coll]
  (when-let [s (seq coll)]
    (lazy-seq
      (let [captured (->> s
                         (drop-to-first before pred-start)
                         (take-to-first before after pred-stop))
            res (drop (count (take-to-first before before pred-start s)) s)]
        (remove nil? (cons captured (partition-inside before after pred-start pred-stop res)))))))

;;;;;;;;;;;;;;
;; File access
(defn reopen [input-filename raf]
  "Close and reopen the provided file.
  This is to be able to follow logrotate when it switches."
  (let [offset (.getFilePointer raf)]
    (.close raf)
    (let [raf (RandomAccessFile. input-filename "r")]
      (if (> offset (.length raf))  ;; detect if the file's been truncated
        (.seek raf 0)
        (.seek raf offset))
      raf)))

(defn raf-seq
  "Tail the provided file."
  [input-filename #^RandomAccessFile raf]
  (if-let [line (.readLine raf)]
    (lazy-seq (cons line (raf-seq input-filename raf)))
    (do (Thread/sleep 1000)
        (recur input-filename
               (reopen input-filename raf)))))

(defn tail-seq
  "Seek the end of the input file and start tailing it."
  [input-filename]
  (let [raf (RandomAccessFile. input-filename "r")]
    (.seek raf (.length raf))
    (raf-seq input-filename raf)))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Traceback "fencepost" predicates

(defn capture
  "Given a regex, exclude or include based on value of negative
  and whether the regex returns a value after being run over s."
  [regex negative s]
  (if-not (nil? s)
    (if negative
      (nil? (re-find regex s))
      (not (nil? (re-find regex s))))
    false))

(def start-capture
  "Start capturing data after this if it returns true."
  (partial capture (start-fencepost) false))

(def end-capture
  "Stop capturing data if we get a line that doesn't match this regex."
  (partial capture (end-fencepost) true))

(defn extract-lines [before after input]
  "Given a streaming input (a log), outputs only what's found between
  and including the 'fencepost' predicates start-capture and end-capture."
  (partition-inside before after start-capture end-capture input))

;;;;;;;;;;;;;;;;;;
;; Subject Builder

(defn traceback-type
  "Extract the traceback type based on regex from a traceback."
  [coll]
  (first (filter #(re-find #"^[a-zA-Z_.]+:\s+" %) coll)))

(defn subject
  "Build the message subject from a prefix and a traceback."
  [s coll]
  (str s " - " (traceback-type coll)))

;;;;;;;;;;;;;;;;;;;;
;; Command & Control
(defn timestamp!
  "Get a human readable timestamp local to the box."
  []
  (format-time/unparse (format-time/formatter "yyyy-MM-dd HH:mm:ss,SSS") (-time/now)))

(defn ->log!
  "Send a message to STDOUT as a log message"
  [sub message]
  (log/info (str "------------------------------------------------------------------\n"
                 (timestamp!) "\n"
                 sub ":\n"
                 message)))

(defn ->sns!
  "Send a message to SNS."
  [arn sub message]
  (sns/publish :topic-arn arn
               :subject sub
               :message (str (timestamp!) "\n" message))
  (log/info (str "Traceback sent to " arn " with subject: " sub)))

(defn data->
  "Based on provided info, grab tracebacks and send them to the endpoint (log or SNS)"
  ([subject-prefix filename before after]  ;; useful for testing purposes, just send tracebacks to log/info
   (data-> ->log! subject-prefix filename before after ""))
  ([subject-prefix filename arn before after]  ;; send tracebacks to the sns topic
   (log/info (str "Sending tracebacks to SNS Topic: " arn
                  ", Subject Prefix: " subject-prefix))
   (data-> (partial ->sns! arn) subject-prefix filename before after ""))
  ([out-fn subject-prefix filename before after _]
   (doseq
     [traceback (extract-lines before after (tail-seq filename))]
     (let [traceback (remove empty? traceback)  ;; tracebacks contain a number of unnecessary newlines
           sub (subject subject-prefix traceback)]
       (out-fn sub (join "\n" traceback))))))


(def cli-options
  [["-a" "--after LINES" "number of lines to capture after the traceback"
    :id :after
    :default 0
    :parse-fn #(Integer/parseInt %)]
   ["-b" "--before LINES" "number of lines to cpature before the traceback"
    :id :before
    :default 0
    :parse-fn #(Integer/parseInt %)]
   ["-h" "--help"]])


(defn -main
  "Entrypoint
  Requires the filename of the log to tail as the first arg."
  [& args]
  (let [topic (slack-sns-topic)
        subject (slack-sns-subject)])

  (let [{:keys [options arguments errors summary]} (parse-opts args cli-options)
        topic (slack-sns-topic)
        subject (slack-sns-subject)]
    (cond
      (:help options) (log/info summary)
      errors (log/error (join "\n" errors))
      (empty? arguments) (log/error summary)
      :else (pmap #(data-> (str subject " (" %1 ")") %1 topic (:before options) (:after options)) arguments))))

;;;;;;;;;;;;;;;;;;;;
;; Helpful Dev Stuff
(comment
  (clojure.pprint/pprint
    (extract-lines 0 0 ["Traceback (most recent call last):"
                        "  File \"./send_status.py\", line 227, in <module>"
                        "smtplib.SMTPDataError: (454, 'Temporary service failure')"
                        ""
                        "Fri Oct 21 12:15:40 UTC 2016 [skipping] report"
                        "[skipping] Running report"]))

  (clojure.pprint/pprint
    (extract-lines 0 0 ["Traceback (most recent call last):"
                        "  File \"./send_status.py\", line 227, in <module>"
                        "smtplib.SMTPDataError: (454, 'Temporary service failure')"
                        ""
                        "Attempt 1 failed! Trying again in 4 seconds..."
                        "Fri Oct 21 12:15:40 UTC 2016 [skipping] report"
                        "[skipping] Running report"]))

  (clojure.pprint/pprint
    (extract-lines 0 0 ["Traceback (most recent call last):"
                        "  File \"./send_status.py\", line 227, in <module>"
                        "smtplib.SMTPDataError: (454, 'Temporary service failure')"
                        ""
                        "Attempt 1 failed and there are no more attempts left!"
                        "Fri Oct 21 12:15:40 UTC 2016 [skipping] report"
                        "[skipping] Running report"]))

  (data-> "testing" "dev-resources/test.log" 2 2))
